### PR TITLE
Run CompositeImplicit decomps when make_fx on the nn module

### PR DESCRIPTION
**Pull request by JackCaoG**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123805
* #123683

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

**Comment by JackCaoG** [comment]
Ah I was debugging https://github.com/pytorch/pytorch/pull/123489#issuecomment-2043915406 where the `fake_mode` does not work, I think it is due to `fake_mode` conflict with the `enable_python_dispatcher` in https://github.com/pytorch/pytorch/pull/123805/files#diff-173d31c139a707aac50fa9b662d552dbf1bff6cb1f207a6565a559f968c13f71R132-R135
```
    with fake_mode and enable_python_dispatcher() and decomp_for_pre_dispatch():
        fx_g = make_fx(wrapped_fn, pre_dispatch=True)(
            fake_value_args, fake_value_kwargs
        )
```

if I remove the `enable_python_dispatcher` or just reordered it to `with enable_python_dispatcher() and fake_mode` I can see `arange` being created as a faketensor.

I am guessing both of them overwrite the `__torch_dispatch__`. I think I need both, @bdhirsh do you know what's correct way to stack these dispatch modes?

**Comment by JackCaoG** [comment]
I am trying to fix `test/test_quantization.py TestQuantizePT2E.test_allow_exported_model_train_eval` and run into some issue. What this test does is to flip the `train`/`eval` option after the model being exported, it is achieved by doing the pattern marching on the FX. Prior to my change. the pattern looks like
match pattern
```
def forward(self, args_0):
    arg0, = fx_pytree.tree_flatten_spec(([args_0], {}), self._in_spec)
    arg0_1 = arg0
    dropout = torch.ops.aten.dropout.default(arg0_1, 0.5, True);  arg0_1 = None
    return pytree.tree_unflatten([dropout], self._out_spec)
```

replace pattern
```
def forward(self, args_0):
    arg0, = fx_pytree.tree_flatten_spec(([args_0], {}), self._in_spec)
    arg0_1 = arg0
    dropout = torch.ops.aten.dropout.default(arg0_1, 0.5, False);  arg0_1 = None
    return pytree.tree_unflatten([dropout], self._out_spec)
```


However after my change to do compositeimplict decomp the pattern changed to 
match pattern
```
def forward(self, args_0):
    arg0, = fx_pytree.tree_flatten_spec(([args_0], {}), self._in_spec)
    arg0_1 = arg0
    native_dropout = torch.ops.aten.native_dropout.default(arg0_1, 0.5, True);  arg0_1 = None
    getitem = native_dropout[0];  native_dropout = None
    return pytree.tree_unflatten([getitem], self._out_spec)
```
replace pattern
```
def forward(self, args_0):
    arg0, = fx_pytree.tree_flatten_spec(([args_0], {}), self._in_spec)
    arg0_1 = arg0
    clone = torch.ops.aten.clone.default(arg0_1);  arg0_1 = None
    return pytree.tree_unflatten([clone], self._out_spec)
```
Note that there isn't a `torch.ops.aten.native_dropout.default(arg0_1, 0.5, Flase);`, but only a clone op. This makes flipping from `eval` to `train` a bit dangerous since it might incorrectly replace a `clone` with `dropout`.

**Comment by JackCaoG** [comment]
Continuing on the last thread, making matter worse, when in place is true, the original pattern looks like
match
```
def forward(self, args_0):
    arg0, = fx_pytree.tree_flatten_spec(([args_0], {}), self._in_spec)
    arg0_1 = arg0
    dropout_ = torch.ops.aten.dropout_.default(arg0_1, 0.5, True);  arg0_1 = None
    return pytree.tree_unflatten([dropout_], self._out_spec)
```
replace
```
def forward(self, args_0):
    arg0, = fx_pytree.tree_flatten_spec(([args_0], {}), self._in_spec)
    arg0_1 = arg0
    dropout_ = torch.ops.aten.dropout_.default(arg0_1, 0.5, False);  arg0_1 = None
    return pytree.tree_unflatten([dropout_], self._out_spec)
```

after my change
match
```
def forward(self, args_0):
    arg0, = fx_pytree.tree_flatten_spec(([args_0], {}), self._in_spec)
    arg0_1 = arg0
    empty_like = torch.ops.aten.empty_like.default(arg0_1, memory_format = torch.contiguous_format)
    bernoulli_ = torch.ops.aten.bernoulli_.float(empty_like);  empty_like = None
    div_ = torch.ops.aten.div_.Scalar(bernoulli_, 0.5);  bernoulli_ = None
    mul_ = torch.ops.aten.mul_.Tensor(arg0_1, div_);  arg0_1 = div_ = None
    return pytree.tree_unflatten([mul_], self._out_spec)
```
replace
```
def forward(self, args_0):
    arg0, = fx_pytree.tree_flatten_spec(([args_0], {}), self._in_spec)
    arg0_1 = arg0
    return pytree.tree_unflatten([arg0_1], self._out_spec)
```

Whoch make sense because in eval model the dropout is a no-op, but this makes pattern matching very tricky...

**Comment by JackCaoG** [comment]
continue torch op decomp in https://github.com/pytorch/pytorch/pull/125808
